### PR TITLE
[CVP-1752] Support operator bundles in CPaaS Midstream

### PR DESCRIPTION
**Pull request by dirgim**
Changes by @samvarankashyap updating midstream image CI and tests:
* Update the image to include dependencies like umoci , skopeo
* Update the input for the CVP image for midstream to accept and process the IMAGE_TO_TEST env variable.
* Run playbooks extract-operator-bundle.yml, validate-operator-bundle.yml as part of the midstream test
* Adding tests in ci using bundles from public container image registry (one positive testcase and 2+ negative testcase)
* Added output of the bundle validation to the validate_operator_bundle role

Fixed yamllint warnings

### CODE REVIEW

**Comment by 14rcole** [comment]
LGTM

**Review by 14rcole — APPROVED** [review]

**Review by jsztuka — APPROVED** [review]
Lgtm

File: .github/workflows/build_and_test_midstream_image.yml [hunk 1]
```diff
@@ -28,31 +27,27 @@ jobs:
...
+      - name: "Run example operator-metadata through midstream ci image  --> negative test"
         continue-on-error: true
         run: |
-          # cd into Dockerfiles since it resets everystep
-          cd Dockerfiles/midstream/
-          mkdir empty-example-bundle
-          podman run -it -v ./empty-example-bundle:/project/operator-bundle -v ./output_logs:/project/output midstream_image:latest
+          rm -rf ./output_logs
+          mkdir ./output_logs
+          podman run -it -v $PWD:/project/operator-test-playbooks -v ./output_logs:/project/output -e IMAGE_TO_TEST=quay.io/somahesh/missing-alm-examples:v0.1.0 midstream_image:latest
```
Comment (sonam1412) [inline .github/workflows/build_and_test_midstream_image.yml:42]: Should we move this test image `quay.io/somahesh/missing-alm-examples:v0.1.0` under quay.io/cvpops or its ok to have it as it is ?
This comment is just a question and has no intention to hold this MR ?

Comment (dirgim) [reply .github/workflows/build_and_test_midstream_image.yml:42]: That's a good point, I'll move the image to the cvpops/test-operator repository

Comment (dirgim) [reply .github/workflows/build_and_test_midstream_image.yml]: @sonam1412 moved the image, please take a look

**Review by sonam1412 — APPROVED** [review]
lgtm

File: Dockerfiles/midstream/Dockerfile [hunk 2]
```diff
@@ -8,8 +8,9 @@ RUN export ARCH=$(case $(arch) in x86_64) echo -n amd64 ;; aarch64) echo -n arm6
...
     && chmod a+x /usr/local/bin/operator-sdk && \
     mkdir /project/output && \
-    dnf install --setopt=install_weak_deps=False -y git-core ansible && \
-    dnf clean all && \
-    git clone --branch $OPERATOR_TEST_PLAYBOOKS_TAG https://github.com/redhat-operator-ecosystem/operator-test-playbooks.git
-COPY ./entrypoint.sh /project/entrypoint.sh
-ENTRYPOINT ["/project/entrypoint.sh"]
+    dnf install --setopt=install_weak_deps=False -y git-core ansible skopeo python3-pip dnf-plugins-core && \
+    dnf copr -y enable ganto/umoci && \
+    dnf install -y umoci && \
```
Comment (lslebodn) [inline Dockerfiles/midstream/Dockerfile:13]: I would rather "install"  umoci from github upstream repo.
Or we should package rpm ourselves. Not sure how reliable is that copr repository

File: Dockerfiles/midstream/Dockerfile [hunk 3]
```diff
@@ -8,8 +8,9 @@ RUN export ARCH=$(case $(arch) in x86_64) echo -n amd64 ;; aarch64) echo -n arm6
     curl -L -o /usr/local/bin/operator-sdk ${OPERATOR_SDK_DL_URL}/operator-sdk_${OS}_${ARCH} \ 
     && chmod a+x /usr/local/bin/operator-sdk && \
     mkdir /project/output && \
-    dnf install --setopt=install_weak_deps=False -y git-core ansible && \
-    dnf clean all && \
-    git clone --branch $OPERATOR_TEST_PLAYBOOKS_TAG https://github.com/redhat-operator-ecosystem/operator-test-playbooks.git
-COPY ./entrypoint.sh /project/entrypoint.sh
-ENTRYPOINT ["/project/entrypoint.sh"]
+    dnf install --setopt=install_weak_deps=False -y git-core ansible skopeo python3-pip dnf-plugins-core && \
```
Comment (lslebodn) [inline Dockerfiles/midstream/Dockerfile:11]: Why do we need to install pip ? We do not install anything with pip in the Dockerfile.
And Dockerfile should ideally contain all necessary dependencies for running tests.
it will reduce random network issues in the execution of tests.

**Comment by lslebodn** [comment]
We should also ask midstream people to review changes.
They might prefer different reporting. 
And it would be better to change it know rather than try to be backward compatible later.

**Review by lslebodn — CHANGES_REQUESTED** [review]
.

File: Dockerfiles/midstream/Dockerfile (hunk 2 above)
Comment (dirgim) [reply Dockerfiles/midstream/Dockerfile]: Changed to fetch the umoci binary from the upstream repository.

File: Dockerfiles/midstream/Dockerfile (hunk 3 above)
Comment (dirgim) [reply Dockerfiles/midstream/Dockerfile]: Removed installation of pip

File: Dockerfiles/midstream/Dockerfile [hunk 4]
```diff
@@ -2,14 +2,16 @@ FROM registry.fedoraproject.org/fedora:33
...
+    chmod a+x /usr/local/bin/operator-sdk && \
+    curl -fL -o /usr/local/bin/umoci https://github.com/opencontainers/umoci/releases/download/${UMOCI_VERSION}/umoci.amd64 && \
+    chmod a+x /usr/local/bin/umoci && \
     mkdir /project/output && \
-    dnf install --setopt=install_weak_deps=False -y git-core ansible && \
-    dnf clean all && \
-    git clone --branch $OPERATOR_TEST_PLAYBOOKS_TAG https://github.com/redhat-operator-ecosystem/operator-test-playbooks.git
-COPY ./entrypoint.sh /project/entrypoint.sh
-ENTRYPOINT ["/project/entrypoint.sh"]
+    dnf install --setopt=install_weak_deps=False -y git-core ansible skopeo dnf-plugins-core && \
```
Comment (lslebodn) [inline Dockerfiles/midstream/Dockerfile:14]: `dnf-plugins-core` was needed just for the dnf copr plugin which we do not need anymore.

File: Dockerfiles/midstream/run_tests.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        result = self.run_validate_operator_bundle()
+        self.assertEqual(0,
+                         result.returncode,
+                         "validate-operator-bundle.yml failed")
+        self.assertTrue(path.exists("/project/output/validation-rc.txt"))
+        self.assertTrue(path.exists("/project/output/validation-output.txt"))
+        self.assertTrue(path.exists("/project/output/validation-version.txt"))
+        with open('/project/output/validation-rc.txt', 'r') as reader:
+            rc = reader.read()
+            self.assertEqual(0, int(rc), "validation of bundle image failed")
```
Comment (lslebodn) [inline Dockerfiles/midstream/run_tests.py:62]: I've just checked logs files from gitlab actions and I realized there is a huge problem with such executions of tests.
We will not display output of validation anywhere. People will not know what to fix.

```
TASK [validate_operator_bundle : Validate the operator bundle manifest and metadata with operator-sdk bundle validate] ***
task path: /project/operator-test-playbooks/roles/validate_operator_bundle/tasks/main.yml:31
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<localhost> EXEC /bin/sh -c 'echo ~root && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1618940583.1714268-409-186136952484899 `" && echo ansible-tmp-1618940583.1714268-409-186136952484899="` echo /root/.ansible/tmp/ansible-tmp-1618940583.1714268-409-186136952484899 `" ) && sleep 0'
Using module file /usr/lib/python3.9/site-packages/ansible/modules/commands/command.py
<localhost> PUT /root/.ansible/tmp/ansible-local-333fve6fbk1/tmpz1tunzmb TO /root/.ansible/tmp/ansible-tmp-1618940583.1714268-409-186136952484899/AnsiballZ_command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1618940583.1714268-409-186136952484899/ /root/.ansible/tmp/ansible-tmp-1618940583.1714268-409-186136952484899/AnsiballZ_command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python3 /root/.ansible/tmp/ansible-tmp-1618940583.1714268-409-186136952484899/AnsiballZ_command.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1618940583.1714268-409-186136952484899/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "cmd": "/usr/local/bin/operator-sdk bundle validate  --verbose /project/test_operator_work_dir/ > /project/output//validation-output.txt 2>&1",
```

We will also need to write another negative test which will fail  just in `validate-operator-bundle.yml`

File: Dockerfiles/midstream/Dockerfile (hunk 4 above)
Comment (dirgim) [reply Dockerfiles/midstream/Dockerfile]: Removed `dnf-plugins-core`

File: Dockerfiles/midstream/run_tests.py (hunk 5 above)
Comment (dirgim) [reply Dockerfiles/midstream/run_tests.py:62]: Added output of the bundle validation to the `validate_operator_bundle` role. Working on a negative test.

Comment (dirgim) [reply Dockerfiles/midstream/run_tests.py:62]: Added the negative test case that fails just in the `validate-operator-bundle.yml` role.

**Comment by dirgim** [comment]
@lslebodn please re-review

File: Dockerfiles/midstream/run_tests.py (hunk 5 above)
Comment (lslebodn) [reply Dockerfiles/midstream/run_tests.py:62]: Thank you for another negative test.

I am not sure whether people will be able to find where is the problem from the output.
But we can refactor the output later. we shoudl unblock midstream folks.

**Review by lslebodn — APPROVED** [review]
lgtm from high level POV

File: Dockerfiles/midstream/run_tests.py (hunk 5 above)
Comment (dirgim) [reply Dockerfiles/midstream/run_tests.py:62]: Ack, I'll create a story to improve the ansible playbook output in general as this is an issue with all of the existing tests.

**Review by nmars — APPROVED** [review]
LGTM
